pkg/crc/api/api.go

```diff
@@ -0,0 +1,88 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net"
+
+	"github.com/code-ready/crc/pkg/crc/logging"
+)
+
+type ArgsType map[string]string
+type handlerFunc func(ArgsType) string
+
+type CrcApiServer struct {
+	listener net.Listener
+	handlers map[string]handlerFunc // relates commands to handler func
+}
+
+// commandRequest struct is used to decode the json request from tray
+type commandRequest struct {
+	Command string            `json:"command"`
+	Args    map[string]string `json:"args,omitempty"`
+}
+
+func CreateApiServer(socketPath string) CrcApiServer {
+	lnr, err := net.Listen("unix", socketPath)
+	if err != nil {
+		logging.Error("Failed to create socket: ", err.Error())
+		return CrcApiServer{}
+	}
+	apiServer := CrcApiServer{
+		listener: lnr,
+		handlers: map[string]handlerFunc{
+			"start":         startHandler,
+			"stop":          stopHandler,
+			"status":        statusHandler,
+			"delete":        deleteHandler,
+			"version":       versionHandler,
+			"webconsoleurl": webconsoleURLHandler,
+		},
+	}
+	return apiServer
+}
+
+func (api CrcApiServer) Serve() {
+	for {
+		conn, err := api.listener.Accept()
+		if err != nil {
+			logging.Error("Error establishing communication: ", err.Error())
```

Comment:
You mean not returned? if an error is encountered its being logged

Reply:
there is no output to the commandline for the daemon. it runs as a background process?

Reply:
> there is no output to the commandline for the daemon. it runs as a background process?

when you run `crc daemon` it blocks and logs the various o/p from the handlers, which runs as go routines.

Reply:
Ah, I missed the `for {}` actually. But we should not be calling `go api.handleConnection(conn)` when an error occurred, which is fixed in your latest iteration.
Is `go api.handleConnection(conn)` asynchronous? If yes, is it possible that we treat requests out of order?
```
conn1, err := api.listener.Accept()
go api.handleConnection(conn1) // asynchronously call api.handleConnection with conn1
conn2, err := api.listener.Accept()
go api.handleConnection(conn2) // asynchronously call api.handleConnection with conn2
```
is there anything which guarantees that the first `go api.handleConnection()` call will complete fully before the second `go api.handleConnection()` call? I know next to nothing on go routines, so maybe it works as expected, maybe not, I have no idea :(

Reply:
No there is no guarantee like that, the requests will be treated out of order. This would cause problems, e.g when a start operation is happening in the go routine and a stop operation is requested, but we prevent this from happening in the tray UI by disabling the menu option.

Reply:
Actually, why use a go routine here, instead of having a blocking call to handleConnection?

Reply:
So that the user can still get the status of the cluster, even when a start/stop/delete action is happening in the background.

Reply:
> is there anything which guarantees that the first go api.handleConnection() call will complete fully before the second go api.handleConnection() call?

I was worried about receiving (for example) a `Failed` answer without having any idea which request it corresponds to. But since we are using a new connection for each request, this kind of confusion actually can't happen.
Running in go routines code which was not meant for that purpose may cause unexpected issues though.